segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -136,52 +137,188 @@ public void initialize(long epoch) {
+            } catch (IOException fnf) {
+                // This can happen if we get a concurrent call to SealOperation with an empty last file; the last file will
+                // be deleted in that case so we need to try our luck again (in which case we need to refresh the file list).
+                if (++attemptCount < MAX_ATTEMPT_COUNT) {
+                    continue;
+                }
+                HDFSExceptionHelpers.throwException(streamSegmentName, fnf);
+            }
+
+            result = StreamSegmentInformation.builder().name(streamSegmentName).length(last.getLen()).sealed(isSealed).build();
+        } while (result == null);
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName, result);
+        return result;
     }
 
     @Override
-    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
-        return call(new CreateOperation(streamSegmentName, this.context));
+    public boolean exists(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+        FileStatus status;
+        try {
+            status = findStatusForSegment(streamSegmentName, false);
+        } catch (IOException e) {
+            return false;
+        }
+        boolean exists = status != null;
+        LoggerHelpers.traceLeave(log, "exists", traceId, streamSegmentName, exists);
+        return exists;
     }
 
-    @Override
-    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
-        return call(new OpenWriteOperation(streamSegmentName, this.context));
+    private boolean isSealed(Path path) throws FileNameFormatException {
+        return getEpocFromPath(path) == MAX_EPOC;
+    }
+
+    protected FileSystem openFileSystem(Configuration conf) throws IOException {
+        return FileSystem.get(conf);
     }
 
     @Override
-    public SegmentHandle openRead(String streamSegmentName) throws StreamSegmentException {
-        return call(new OpenReadOperation(streamSegmentName, this.context));
+    public int read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle, offset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0 || buffer.length < bufferOffset + length) {
+            throw new ArrayIndexOutOfBoundsException(String.format(
+                    "Offset (%s) must be non-negative, and bufferOffset (%s) and length (%s) must be valid indices into buffer of size %s.",
+                    offset, bufferOffset, length, buffer.length));
+        }
+
+        Timer timer = new Timer();
+
+        int attemptCount = 0;
+        AtomicInteger totalBytesRead = new AtomicInteger();
+        boolean needsRefresh = true;
+        while (needsRefresh && attemptCount < MAX_ATTEMPT_COUNT) {
```

Comment:
This is the same retry pattern as above. If we can do this with RetryHelper we should. If not, we should figure out what is missing so we can generalize the solution.